great_expectations/core/async_executor.py

```diff
@@ -0,0 +1,133 @@
+
+    WARNING: This class is experimental.
+    """
+
+    def __init__(self, future: Future = None, value: Any = None):
+        """AsyncResult instances are created by AsyncExecutor.submit() and should not otherwise be created directly."""
+        self._future = future
+        self._value = value
+
+    def result(self):
+        """Return the value corresponding to the AsyncExecutor.submit() call, blocking if necessary until the execution
+        finishes.
+        """
+        return self._future.result() if self._future is not None else self._value
+
+
+class AsyncExecutor:
+    """Wrapper around ThreadPoolExecutor to facilitate single code path
+    for both when concurrency is enabled and disabled.
+
+    WARNING: This class is experimental.
+    """
+
+    def __init__(
+        self,
+        concurrency_config: ConcurrencyConfig,
+        max_workers: int,
+    ):
+        """Initializes a new AsyncExecutor instance used to organize code for multithreaded execution.
+
+        If multithreading is disabled, all execution will be done synchronously (e.g. on the main thread) during the
+        call to submit. This is useful for introducing multithreading while still supporting single threaded execution.
+        This allows configuration to determine whether or not multithreading is used, which is useful when there are
+        concerns that some workflows may not benefit from multithreading or may not be safe to multithread.
+
+        This class is intended to be used as a context manager using the `with` statement.
+
+        Args:
+            concurrency_config: Configuration used to determine whether or not execution is done concurrently with
+                multiple threads. Even if the configuration has concurrency enabled, if max_workers is 1 then all work
+                will be done synchronously (e.g. on the main thread) during the call to submit.
+            max_workers: The maximum number of threads that can be used to execute concurrently. If concurrency is
+                disabled or max_workers is 1, all work will be done synchronously (e.g. on the main thread) during the
+                call to submit. Note that the maximum number of threads is also limited by
+                concurrency_config.max_database_query_concurrency.
+        """
+        # Only enable concurrent execution if it is enabled in the config AND there is more than 1 max worker specified.
+        self._execute_concurrently = concurrency_config.enabled and max_workers > 1
+
+        self._thread_pool_executor = (
+            ThreadPoolExecutor(
+                max_workers=min(
+                    concurrency_config.max_database_query_concurrency,
+                    max_workers,
+                )
+            )
+            if self._execute_concurrently
+            else None
+        )
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, ext, exv, trb):
+        self.shutdown()
```

Comment:
Feedback from Evan: add in docstring what is being done or not done with the exception, and if this is an OK outcome. Are these automatically propagated up?

Reply:
Thanks @eap for the review feedback! This is fixed in latest commit a1897c7258f18ed318efea430a0bafb05422f5b3